lib/active_model/serializer/collection_serializer.rb

```diff
@@ -73,7 +80,8 @@ def serializers_from_resources
       def serializer_from_resource(resource, serializer_context_class, options)
         serializer_class = options.fetch(:serializer) { serializer_context_class.serializer_for(resource) }
 
-        if serializer_class.nil? # rubocop:disable Style/GuardClause
+        if serializer_class.nil?
+          yield if block_given?
```

Comment:
For this pattern to work, it must either yield or raise.  As written it will yield then raise

```ruby
block_given? and yield or
           fail NoSerializerError, "No serializer found for resource: #{resource.inspect}"
```

would work, I think (not sure about the `and`)